README.md
<h1 align="center">parquet-go</h1>
<p align="center">
        <a href="https://github.com/fraugster/parquet-go/releases"><img src="https://img.shields.io/github/v/tag/fraugster/parquet-go.svg?color=brightgreen&label=version&sort=semver"></a>
        <a href="https://circleci.com/gh/fraugster/parquet-go/tree/master"><img src="https://circleci.com/gh/fraugster/parquet-go/tree/master.svg?style=shield"></a>
        <a href="https://goreportcard.com/report/github.com/fraugster/parquet-go"><img src="https://goreportcard.com/badge/github.com/fraugster/parquet-go"></a>
        <a href="https://codecov.io/gh/fraugster/parquet-go"><img src="https://codecov.io/gh/fraugster/parquet-go/branch/master/graph/badge.svg"/></a>
        <a href="https://godoc.org/github.com/fraugster/parquet-go"><img src="https://img.shields.io/badge/godoc-reference-blue.svg?color=blue"></a>
        <a href="https://github.com/fraugster/parquet-go/blob/master/LICENSE"><img src="https://img.shields.io/badge/license-Apache%202-blue"></a>
</p>

---

`parquet-go` is an implementation of the [Apache Parquet file format](https://github.com/apache/parquet-format)
in Go. It provides functionality to both read and write Parquet files, as well
as high-level functionality to manage the data schema of Parquet files, to
directly write Go objects to Parquet files using automatic or custom
marshalling and to read records from Parquet files into Go objects using
automatic or custom marshalling.

Parquet is a file format to store nested data structures in a flat columnar
format. By storing in a column-oriented way, it allows for efficient reading
of individual columns without having to read and decode complete rows. This
allows for efficient reading and faster processing when using the file format
in conjunction with distributed data processing frameworks like Apache Hadoop
or distributed SQL query engines like Presto and AWS Athena.

This implementation is divided into several packages. The top-level package is
the low-level implementation of the Parquet file format. It is accompanied by
the sub-packages `parquetschema` and `floor`. `parquetschema` provides
functionality to parse textual schema definitions as well as the data types to
manually or programmatically construct schema definitions. `floor` is a
high-level wrapper around the low-level package. It provides functionality to
open Parquet files to read from them or write to them using automated or custom
marshalling and unmarshalling.

## Supported Features

| Feature                                  | Read | Write | Note |
| ---------------------------------------- | ---- | ----- | ---- |
| Compression                              | Yes  | Yes   | Only GZIP and SNAPPY are supported out of the box, but it is possible to add other compressors, see below. |
| Dictionary Encoding                      | Yes  | Yes   |
| Run Length Encoding / Bit-Packing Hybrid | Yes  | Yes   | The reader can read RLE/Bit-pack encoding, but the writer only uses bit-packing. |
| Delta Encoding                           | Yes  | Yes   |
| Byte Stream Split                        | No   | No    |
| Data page V1                             | Yes  | Yes   |
| Data page V2                             | Yes  | Yes   |
| Statistics in page meta data             | No   | Yes   | Page meta data is generally not made available to users and not used by `parquet-go`. |
| Index Pages                              | No   | No    |
| Dictionary Pages                         | Yes  | Yes   |
| Encryption                               | No   | No    |
| Bloom Filter                             | No   | No    |
| Logical Types                            | Yes  | Yes   | Support for logical type is in the high-level package `floor` the low level Parquet library only supports the basic types, see the type mapping table. |

## Supported Data Types

| Parquet Type            | Go Type         | Note |
| ----------------------- | --------------- | ---- |
| BOOLEAN                 | bool            |
| INT32                   | int32           | See the note about the int type. |
| INT64                   | int64           | See the note about the int type. |
| INT96                   | [12]byte        |
| FLOAT                   | float32         |
| DOUBLE                  | float64         |
| BYTE\_ARRAY             | string, []byte  |
| FIXED\_LEN\_BYTE\_ARRAY | []byte, [N]byte |

Note: the low-level implementation only supports int32 for the INT32 type and int64 for the INT64 type in Parquet.
Plain int or uint are not supported. The high-level `floor` package contains more extensive support for these
data types.

## Supported Logical Types

| Logical Type | Mapped to Go types           | Note |
| ------------ | -----------------------      | ---- |
| STRING       | string, []byte               |
| DATE         | int32, time.Time             | int32: Days since Unix epoch (Jan 01 1970 00:00:00 UTC); time.Time: only in `floor` |
| TIME         | int32, int64, goparquet.Time | int32: TIME(MILLIS, ...), int64: TIME(MICROS, ...), TIME(NANOS, ...); goparquet.Time: only in `floor` |
| TIMESTAMP    | int64, [12]byte, time.Time   | time.Time: only in `floor` |
| UUID         | [16]byte                     |
| LIST         | []T, [N]T                    | Slices and arrays of any type. |
| MAP          | map[T1]T2                    | Maps with any key and value types. |
| ENUM         | string, []byte               |
| JSON         | string, []byte               |
| BSON         | string, []byte               |
| DECIMAL      | int32, int64,[]byte, [N]byte |
| INTEGER      | {,u}int{,8,16,32,64}         | Implementation is loose and will allow any INTEGER logical type converted to any signed or unsigned int Go type. |

## Supported Converted Types

| Converted Type        | Mapped to Go types  | Note |
| --------------------  | ------------------- | ---- |
| UTF8                  | string, []byte      |
| TIME\_MILLIS          | int32               | Number of milliseconds since the beginning of the day. |
| TIME\_MICROS          | int64               | Number of microseconds since the beginning of the day. |
| TIMESTAMP\_MILLIS     | int64               | Number of milliseconds since Unix epoch (Jan 01 1970 00:00:00 UTC). |
| TIMESTAMP\_MICROS     | int64               | Number of milliseconds since Unix epoch (Jan 01 1970 00:00:00 UTC). |
| {,U}INT\_{8,16,32,64} | {,u}int{8,16,32,64} | Implementation is loose and will allow any converted type with any int Go type. |
| INTERVAL              | [12]byte            |

Please note that converted types are deprecated. Logical types should be used preferably.

## Supported Compression Algorithms

| Compression Algorithm | Supported | Notes |
| --------------------- | --------- | ----- |
| GZIP                  | Yes; Out of the box |
| SNAPPY                | Yes; Out of the box |
| BROTLI                | Yes; By importing [github.com/akrennmair/parquet-go-brotli](https://github.com/akrennmair/parquet-go-brotli) |
| LZ4                   | No | LZ4 has been deprecated as of parquet-format 2.9.0. |
| LZ4\_RAW              | Yes; By importing [github.com/akrennmair/parquet-go-lz4raw](https://github.com/akrennmair/parquet-go-lz4raw) |
| LZO                   | Yes; By importing [github.com/akrennmair/parquet-go-lzo](https://github.com/akrennmair/parquet-go-lzo) | Uses a cgo wrapper around the original LZO implementation which is licensed as GPLv2+. |
| ZSTD                  | Yes; By importing [github.com/akrennmair/parquet-go-zstd](https://github.com/akrennmair/parquet-go-zstd) |

## Schema Definition

`parquet-go` comes with support for both textual and automatic object schema
definitions.

### Textual Schema Definitions

The sub-package `parquetschema` comes with a parser to turn the textual schema
definition into the right data type to use elsewhere to specify Parquet
schemas. The syntax has been mostly reverse-engineered from a similar format
also supported but barely documented in [Parquet's Java implementation](https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/schema/MessageTypeParser.java).

For the full syntax, please have a look at the [parquetschema package Go documentation](http://godoc.org/github.com/fraugster/parquet-go/parquetschema).

Generally, the schema definition describes the structure of a message. Parquet
will then flatten this into a purely column-based structure when writing the
actual data to Parquet files.

A message consists of a number of fields. Each field either has type or is a
group. A group itself consists of a number of fields, which in turn can have
either a type or are a group themselves. This allows for theoretically
unlimited levels of hierarchy.

Each field has a repetition type, describing whether a field is required (i.e.
a value has to be present), optional (i.e. a value can be present but doesn't
have to be) or repeated (i.e. zero or more values can be present). Optionally,
each field (including groups) have an annotation, which contains a logical type
or converted type that annotates something about the general structure at this
point, e.g. `LIST` indicates a more complex list structure, or `MAP` a key-value
map structure, both following certain conventions. Optionally, a typed field
can also have a numeric field ID. The field ID has no purpose intrinsic to the
Parquet file format.

Here is a simple example of a message with a few typed fields:

```
message coordinates {
    required float64 latitude;
    required float64 longitude;
    optional int32 elevation = 1;
    optional binary comment (STRING);
}
```

In this example, we have a message with four typed fields, two of them
required, and two of them optional. `float64`, `int32` and `binary` describe
the fundamental data type of the field, while `longitude`, `latitude`,
`elevation` and `comment` are the field names. The parentheses contain
an annotation `STRING` which indicates that the field is a string, encoded
as binary data, i.e. a byte array. The field `elevation` also has a field
ID of `1`, indicated as numeric literal and separated from the field name
by the equal sign `=`.

In the following example, we will introduce a plain group as well as two
nested groups annotated with logical types to indicate certain data structures:

```
message transaction {
    required fixed_len_byte_array(16) txn_id (UUID);
    required int32 amount;
    required int96 txn_ts;
    optional group attributes {
        optional int64 shop_id;
        optional binary country_code (STRING);
        optional binary postcode (STRING);
    }
    required group items (LIST) {
        repeated group list {
            required int64 item_id;
            optional binary name (STRING);
        }
    }
    optional group user_attributes (MAP) {
        repeated group key_value {
            required binary key (STRING);
            required binary value (STRING);
        }
    }
}
```

In this example, we see a number of top-level fields, some of which are
groups. The first group is simply a group of typed fields, named `attributes`.

The second group, `items` is annotated to be a `LIST` and in turn contains a
`repeated group list`, which in turn contains a number of typed fields. When
a group is annotated as `LIST`, it needs to follow a particular convention:
it has to contain a `repeated group` named `list`. Inside this group, any
fields can be present.

The third group, `user_attributes` is annotated as `MAP`. Similar to `LIST`,
it follows some conventions. In particular, it has to contain only a single
`required group` with the name `key_value`, which in turn contains exactly two
fields, one named `key`, the other named `value`. This represents a map
structure in which each key is associated with one value.

### Object Schema Definitions

The sub-package `parquetschema/autoschema` supports auto-generating schema
definitions for a provided object's type using reflection and struct tags. The
generated schema is meant to be compatible with the reflection-based
marshalling/unmarshalling in the `floor`.

#### Supported Parquet Types

| Parquet Type            | Go Types                     | Note |
| ----------------------- | ---------------------------- | ---- |
| BOOLEAN                 | bool                         |
| INT32                   | int{8,16,32}, uint{,8,16,32} | 
| INT64                   | int{,64}, uint64             |
| INT96                   | [12]byte                     | Must specify `type=INT96` in the `parquet` struct tag. |
| FLOAT                   | float32                      |
| DOUBLE                  | float64                      |
| BYTE\_ARRAY             | string, []byte               |
| FIXED\_LEN\_BYTE\_ARRAY | []byte, [N]byte              |

#### Supported Logical Types

| Logical Type | Go Types                      | Note |
| ------------ | ----------------------------- | ---- |
| STRING       | string, []byte                |
| MAP          | map[T1]T2                     | Maps with any key and value types. |
| LIST         | []T, [N]T                     | Slices and arrays of any type except for byte. |
| ENUM         | string, []byte                |
| DECIMAL      | int32, int64, []byte, [N]byte |
| DATE         | int32, time.Time              |
| TIME         | int32, int64, goparquet.Time  | int32: TIME(MILLIS, {false,true}), int64: TIME({MICROS,NANOS}, {false,true}) |
| TIMESTAMP    | int64, time.Time              |
| INTEGER      | {,u}int{,8,16,32,64}          |
| JSON         | string, []byte                |
| BSON         | string, []byte                |
| UUID         | [16]byte                      |

Pointers are automatically mapped to optional fields. Unsupported Go types
include funcs, interfaces, unsafe pointers, unsigned int pointers, and complex
numbers.

#### Default Type Mappings

By default, Go types are mapped to Parquet types and in some cases logical
types as well. More specific mappings can be achieved by the use of struct
tags (see below).

| Go Type           | Default Parquet Type    | Default Logical Type |
| ------------------| ----------------------- | -------------------- |
| bool              | BOOLEAN                 |
| int{,8,16,32,64}  | INT{64,32,32,32,64}     | INTEGER({64,8,16,32,64}, true) |
| uint{,8,16,32,64} | INT{32,32,32,32,64}     | INTEGER({32,8,16,32,64}, false) |
| string            | BYTE\_ARRAY             | STRING |
| []byte            | BYTE\_ARRAY             |
| [N]byte           | FIXED\_LEN\_BYTE\_ARRAY |
| time.Time         | INT64                   | TIMESTAMP(NANOS, true)
| goparquet.Time    | INT64                   | TIME(NANOS, true)
| map               | group                   | MAP |
| slice, array      | group                   | LIST |
| struct            | group                   |


#### Struct Tags

Automatic schema definition generation supports the use of the `parquet` struct
tag for further schema specification beyond the default mappings. Tag fields
have the format `key=value` and are comma separated. The tags do not support
converted types as these are now deprecated by Parquet. Since converted types
are still required to support backward compatibility, they are automatically
set based on a field's logical type.

| Tag Field       | Type   | Values                                                                           | Notes |
| --------------- | ------ | -------------------------------------------------------------------------------- | ----- |
| name            | string | ANY                                                                              | Defaults to the lower-case struct field name. |
| type            | string | `INT96`                                                                          | Unless using a [12]byte field for INT96, this does not ever need to be specified. |
| logicaltype     | string | `STRING`, `ENUM`, `DECIMAL`, `DATE`, `TIME`, `TIMESTAMP`, `JSON`, `BSON`, `UUID` | Maps and non-byte slices and arrays are always mapped to MAP and LIST logical types, respectively. |
| timeunit        | string | `MILLIS`, `MICROS`, `NANOS`                                                      | Only used when the logical type is TIME or TIMESTAMP, defaults to `NANOS`. |
| isadjustedtoutc | bool   | ANY                                                                              | Only used when the logical type is TIME or TIMESTAMP, defaults to `true`. |
| scale           | int32  | N >= 0                                                                           | Only used when the logical type is DECIMAL, defaults to 0. |
| precision       | int32  | N >= 0                                                                           | Only used when the logical type is DECIMAL, required. |

All fields must be prefixed by `key.` and `value.` when referring to key and
value types of a map, respectively, and `element.` when referring to the
element type of a slice or array. It is invalid to prefix `name` since it can
only apply to the field itself.

#### Object Schema Example

```
type example  struct {
        ByteSlice          []byte
        String             string
        ByteString         []byte          `parquet:"name=byte_string, logicaltype=STRING"`
        Int64              int64           `parquet:"name=int_64"`
        Uint8              uint8           `parquet:"name=u_int_8"`
        Int96              [12]byte        `parquet:"name=int_96, type=INT96"`
        DefaultTS          time.Time       `parquet:"name=default_ts"`
        Timestamp          int64           `parquet:"name=ts, logicaltype=TIMESTAMP, timeunit=MILLIS, isadjustedtoutc=false`
        Date               time.Time       `parquet:"name=date, logicaltype=DATE"`
        OptionalDecimal    *int32          `parquet:"name=decimal, logicaltype=DECIMAL, scale=5, precision=10"`
        TimeList           []int32         `parquet:"name=time_list, element.logicaltype=TIME, element.timeunit=MILLIS"`
	DecimalTimeMap     map[int64]int32 `parquet:"name=decimal_time_map, key.logicaltype=DECIMAL, key.scale=5, key.precision=15, value.logicaltype=TIME, value.timeunit=MILLIS", value.isadjustedtoutc=true`
        Struct             struct {
                OptionalInt64 *int64   `parquet:"name=int_64"`
	        Time          int64    `parquet:"name=time, logicaltype=TIME, isadjustedtoutc=false"`
	        StringList    []string `parquet:"name=string_list"`
        } `parquet:"name=struct"`
}
```

The above struct is equivalent to the following schema definition:

```
message autogen_schema {
    required binary byteslice;
    required binary string (STRING);
    required binary byte_string (STRING);
    required int64 int_64 (INTEGER(64,true));
    required int32 int_8 (INTEGER(8,false));
    required int96 int_96;
    required int64 default_ts (TIMESTAMP(NANOS,true));
    required int64 ts (TIMESTAMP(MILLIS,false));
    required int32 date (DATE);
    optional int32 decimal (DECIMAL(10,5));
    required group time_list (LIST) {
        repeated group list {
          required int32 element (TIME(MILLIS,true));
        }
    }
    optional group decimal_time_map (MAP) {
        repeated group key_value (MAP_KEY_VALUE) {
          required int64 key (DECIMAL(15,5));
          required int32 value (TIME(MILLIS, true));
        }
    }
    required group struct {
        optional int64 int_64 (INTEGER(64,true));

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Fixed CHANGELOG for v0.11.0.
- Fixed bug in `goparquet.recursiveFix` that was overwriting column paths for repeated group elements in a schema defintion.
- Added richer struct tag support for auto-generated schema definitions.

## [v0.11.0] - 2022-04-21

- Cleaned up API by removing Schema\* interfaces, deprecating function using dotted notation, and introducing `ColumnPath` type instead.
- Improved error handling around row groups.
- Improved write performance.
- Fixed thrift crash bugs.

## [v0.10.0] - 2022-02-18

- Updated to parquet-format 2.9.0.
- Added support for page data statistics.
- Improved test coverage.
- Fixed issue with dictionary encoding where bit width was determined incorrectly.
- Added support for writing per-page CRC32 checksums.
- Added support for optional CRC32 checksum validation on read.
- Fixed issue with float32/64 dictionary encoding where trying to encode NaN failed with an error.

## [v0.9.0] - 2022-02-10

- Implemented lazy loading of pages.
- Added support for writing multiple data pages per row group.

## [v0.8.0] - 2022-02-02
- Set correct total size and total compressed in row group data.
- Fixed delta bit-pack encoding of int32 and int64 and delta-length bit-packing encoding of byte\_arrays when only a single value is written.
- Fixed build issue in the floor package that somehow made it into the previous release.

## [v0.7.0] - 2022-02-01
- Relax schema parser requirement that field names begin with a letter or underscore.
- Remove unsigned support from all integer encodings.
- Introduced explicit Clone method for parquetschema.SchemaDefinition.
- Corrected parsing of DECIMAL as ConvertedType if scale and precision aren't available.
- Fixed nested optional field reads.

## [v0.6.1] - 2021-11-19
- perf - cache result of GetSchemaDefinition call
- updated readme, added special mentions section

## [v0.6.0] - 2021-11-3
- Added a schema generator which uses reflection to automatically generate a parquet schema from a go object.
- Upgraded CI golang to 1.17.2
- Added test to reproduce issue 41: https://github.com/fraugster/parquet-go/issues/41 

## [v0.5.0] - 2021-10-29
- Upgraded thrift to v0.15.0
- Fixed reading & writing of INT96 time

## [v0.4.0] - 2021-10-06
- Fixed issues where fields in structs that were not defined in the schema raised errors (array, slice, time, map)
- Added support for reflect encoding/decoding of additional types to/from parquet types
    - Go type <-> Parquet type
    - int     <-> int64
    - int32   <-> int64
    - int16   <-> int64
    - int8    <-> int64
    - uint    <-> int64
    - uint16  <-> int64
    - uint8   <-> int64
    - int64   <-> int32
    - uint64  <-> int32
    - uint32  <-> int32
- Removed some inconsistent api behaviors
    - reflect marshaling/unmarshalling now ignores fields not defined in the schema (this already happens when
      the marshaller is used by floor.Writer and goparquet.FileReader)
    - int32PlainDecoder and int32PlainEncoder now no longer support uint values
- Added function `IsAfterUnixEpoch` to allow test whether a timestamp can be written as Julian date.
- Allowed setting selected columns after opening a parquet file through `SetSelectedColumns` method.
- Fixed `dictStore` to correctly reset `valueSize` which should reduce size of written files.
- Fixed reflection-based marshal/unmarshaling for some basic types.
- Fixed `DECIMAL`s computation of maximum amount of digits.
- Improved parsing of legacy timestamps.
- Allowed reading of file metadata before properly opening a parquet file.
- Added method `SeekToRowGroup` to allow seeking to specific row groups.

## [v0.3.0] - 2020-12-15
- Added examples how to use the low-level and high-level APIs.
- Added backwards compability in floor reader and writer for lists as produced by Athena.
- Fixed many crash issues found during fuzzing of the parquet reader.

## [v0.2.1] - 2020-11-04
- Release to correct missing changelog.

## [v0.2.0] - 2020-11-04
- Added `csv2parquet` command to convert CSV files into parquet files
- Fixed issue in `parquet-tool cat` that wouldn't print any records if no limit was provided
- Added support for backward compatible MAPs and LISTs in schema validation
- Added ValidateStrict method to validate without regard for backwards compatibility

## [v0.1.1] - 2020-05-26
- Added high-level interface to access file and column metadata

## [v0.1.0] - 2020-04-24
- Initial release

[Unreleased]: https://github.com/fraugster/parquet-go/compare/v0.11.0...HEAD
[v0.11.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.11.0
[v0.10.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.10.0
[v0.9.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.9.0
[v0.8.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.8.0
[v0.7.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.7.0
[v0.6.1]: https://github.com/fraugster/parquet-go/releases/tag/v0.6.1
[v0.6.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.6.0
[v0.5.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.5.0
[v0.4.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.4.0
[v0.3.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.3.0
[v0.2.1]: https://github.com/fraugster/parquet-go/releases/tag/v0.2.1
[v0.2.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.2.0
[v0.1.1]: https://github.com/fraugster/parquet-go/releases/tag/v0.1.1
[v0.1.0]: https://github.com/fraugster/parquet-go/releases/tag/v0.1.0


parquetschema/autoschema/tags.go
package autoschema

import (
	"reflect"
	"strconv"
	"strings"
	"time"

	goparquet "github.com/fraugster/parquet-go"
	"github.com/fraugster/parquet-go/parquet"
	"github.com/fraugster/parquet-go/parquetschema"
	"github.com/pkg/errors"
)

const (
	structTagName            = "name"
	structTagType            = "type"
	structTagLogicalType     = "logicaltype"
	structTagIsAdjustedToUTC = "isadjustedtoutc"
	structTagTimeUnit        = "timeunit"
	structTagScale           = "scale"
	structTagPrecision       = "precision"
	structTagKeyPrefix       = "key."
	structTagValuePrefix     = "value."
	structTagElementPrefix   = "element."
)

func parseParquetTag(field reflect.StructField, columnType reflect.Type, column *parquetschema.ColumnDefinition) error {
	element := column.SchemaElement
	tagFieldMap, err := createTagFieldMap(field, getParquetTagPrefix(element.Name))
	if err != nil {
		return errors.Wrap(err, "creating struct tag field map")
	}

	if name, ok := tagFieldMap[structTagName]; ok {
		element.Name = name
	} else if element.Name == "" {
		element.Name = strings.ToLower(field.Name)
	}

	for len(column.Children) > 0 {
		// This is a column definition with children, just set the name
		// and return since any other struct tag fields are for the
		// children column definitions.
		return nil
	}

	if typeString, ok := tagFieldMap[structTagType]; ok {
		element.Type, err = typeFromString(typeString)
		if err != nil {
			return errors.Wrap(err, "getting type from string")
		}
	}

	if logicalTypeString, ok := tagFieldMap[structTagLogicalType]; ok {
		element.LogicalType, element.ConvertedType, err = logicalTypeFromString(logicalTypeString)
		if err != nil {
			return errors.Wrap(err, "getting the logical type from string")
		}

		if element.LogicalType.DATE != nil {
			// Ensure that the Parquet type is set to INT32 for
			// logical DATE fields since they may have been
			// converted from a time.Time struct field (which
			// defaults to int64).
			element.Type = parquet.TypePtr(parquet.Type_INT32)
		}
	}

	if isAdjustedToUTCString, ok := tagFieldMap[structTagIsAdjustedToUTC]; ok || (element.LogicalType != nil && (element.LogicalType.TIME != nil || element.LogicalType.TIMESTAMP != nil)) {
		isAdjustedToUTC, err := isAdjustedToUTCFromString(isAdjustedToUTCString)
		if err != nil {
			return errors.Wrap(err, "getting is adjusted to UTC from string")
		}
		if element.LogicalType == nil {
			return errors.New("must specify a logical type when specifying is adjusted to UTC")
		}

		if element.LogicalType.TIME != nil {
			element.LogicalType.TIME.IsAdjustedToUTC = isAdjustedToUTC
		} else if element.LogicalType.TIMESTAMP != nil {
			element.LogicalType.TIMESTAMP.IsAdjustedToUTC = isAdjustedToUTC
		} else {
			return errors.Errorf("specifying is adjusted to UTC is incompatible with %s", parquetschema.GetSchemaLogicalType(element.LogicalType))
		}
	}

	if timeUnitString, ok := tagFieldMap[structTagTimeUnit]; ok || (element.LogicalType != nil && (element.LogicalType.TIME != nil || element.LogicalType.TIMESTAMP != nil)) {
		tu, err := timeUnitFromString(timeUnitString)
		if err != nil {
			return errors.Wrap(err, "getting time unit from string")
		}
		if element.LogicalType == nil {
			return errors.New("must specify a logical type when specifying a time unit")
		}

		if element.LogicalType.TIME != nil {
			element.LogicalType.TIME.Unit = tu
			if tu.MILLIS != nil {
				element.ConvertedType = parquet.ConvertedTypePtr(parquet.ConvertedType_TIME_MILLIS)
			} else if tu.MICROS != nil {
				element.ConvertedType = parquet.ConvertedTypePtr(parquet.ConvertedType_TIME_MICROS)
			} else {
				element.ConvertedType = nil
			}
		} else if element.LogicalType.TIMESTAMP != nil {
			element.LogicalType.TIMESTAMP.Unit = tu
			if tu.MILLIS != nil {
				element.ConvertedType = parquet.ConvertedTypePtr(parquet.ConvertedType_TIMESTAMP_MILLIS)
			} else if tu.MICROS != nil {
				element.ConvertedType = parquet.ConvertedTypePtr(parquet.ConvertedType_TIMESTAMP_MICROS)
			} else {
				element.ConvertedType = nil
			}
		} else {
			return errors.Errorf("specifying a time unit is incompatible with %s", parquetschema.GetSchemaLogicalType(element.LogicalType))
		}
	}

	if scaleString, ok := tagFieldMap[structTagScale]; ok {
		scale, err := strconv.ParseInt(scaleString, 10, 32)
		if err != nil {
			return errors.Errorf("converting the specified scale value '%s' to int32", scaleString)
		}
		if element.LogicalType == nil {
			return errors.New("must specify a logical type when specifying scale")
		}
		if element.LogicalType.DECIMAL == nil {
			return errors.Errorf("specifying scale is incompatible with logical type %s", parquetschema.GetSchemaLogicalType(element.LogicalType))
		}

		element.LogicalType.DECIMAL.Scale = int32(scale)
	}

	if precisionString, ok := tagFieldMap[structTagPrecision]; ok {
		precision, err := strconv.ParseInt(precisionString, 10, 32)
		if err != nil {
			return errors.Errorf("converting the specified precision value '%s' to int32", precisionString)
		}
		if element.LogicalType == nil {
			return errors.New("must specify a logical type when specifying precision")
		}
		if element.LogicalType.DECIMAL == nil {
			return errors.Errorf("specifying precision is incompatible with %s", parquetschema.GetSchemaLogicalType(element.LogicalType))
		}

		element.LogicalType.DECIMAL.Precision = int32(precision)
	}

	// Check type compatibility at the end since there may have been some
	// additional logical after setting the Parquet type and Parquet
	// logical type that affects validity (e.g. time unit specification).
	if _, ok := tagFieldMap[structTagType]; ok && !isTypeCompatible(element.Type, columnType) {
		return errors.Errorf("incompatible Go type %s and Parquet type %s", columnType, element.Type.String())
	}
	if _, ok := tagFieldMap[structTagLogicalType]; ok && !isLogicalTypeCompatible(element.LogicalType, columnType) {
		return errors.Errorf("incompatible Go type %s and Parquet logical type %s", columnType, parquetschema.GetSchemaLogicalType(element.LogicalType))
	}

	return nil
}

func getParquetTagPrefix(name string) string {
	switch name {
	case "key":
		return structTagKeyPrefix
	case "value":
		return structTagValuePrefix
	case "element":
		return structTagElementPrefix
	default:
		return ""
	}
}

func createTagFieldMap(field reflect.StructField, prefix string) (map[string]string, error) {
	tagFieldMap := map[string]string{}
	for _, tagField := range strings.Split(field.Tag.Get("parquet"), ",") {
		splitField := strings.Split(tagField, "=")
		if len(splitField) != 2 {
			// The struct tag field does not follow the `key=val`
			// convention, skip it.
			continue
		}

		prefixedKey := strings.TrimSpace(splitField[0])
		if !strings.HasPrefix(prefixedKey, prefix) {
			// Ignore required prefix for this mapping, such as
			// "key." or "value.".
			continue
		}
		key := prefixedKey[len(prefix):]
		if _, ok := tagFieldMap[key]; ok {
			return nil, errors.Errorf("struct tag field '%s' specified more than once", prefixedKey)
		}
		tagFieldMap[key] = strings.TrimSpace(splitField[1])
	}

	return tagFieldMap, nil
}

func typeFromString(s string) (*parquet.Type, error) {
	var t parquet.Type
	switch s {
	case parquet.Type_INT96.String():
		t = parquet.Type_INT96
	default:
		return nil, errors.Errorf("unsupported type '%s' specified", s)
	}

	return parquet.TypePtr(t), nil
}

func logicalTypeFromString(s string) (*parquet.LogicalType, *parquet.ConvertedType, error) {
	var ct *parquet.ConvertedType
	lt := parquet.NewLogicalType()

	switch s {
	case "STRING":
		lt.STRING = parquet.NewStringType()
		ct = parquet.ConvertedTypePtr(parquet.ConvertedType_UTF8)
	case "ENUM":
		lt.ENUM = parquet.NewEnumType()
		ct = parquet.ConvertedTypePtr(parquet.ConvertedType_ENUM)
	case "DECIMAL":
		lt.DECIMAL = parquet.NewDecimalType()
		ct = parquet.ConvertedTypePtr(parquet.ConvertedType_DECIMAL)
	case "DATE":
		lt.DATE = parquet.NewDateType()
		ct = parquet.ConvertedTypePtr(parquet.ConvertedType_DATE)
	case "TIME":
		lt.TIME = parquet.NewTimeType()
	case "TIMESTAMP":
		lt.TIMESTAMP = parquet.NewTimestampType()
	case "JSON":
		lt.JSON = parquet.NewJsonType()
		ct = parquet.ConvertedTypePtr(parquet.ConvertedType_JSON)
	case "BSON":
		lt.BSON = parquet.NewBsonType()
		ct = parquet.ConvertedTypePtr(parquet.ConvertedType_BSON)
	case "UUID":
		lt.UUID = parquet.NewUUIDType()
	default:
		return nil, nil, errors.Errorf("unsupported logical type '%s' specified", s)
	}

	return lt, ct, nil
}

func isAdjustedToUTCFromString(s string) (bool, error) {
	switch strings.ToLower(s) {
	case "true", "":
		return true, nil
	case "false":
		return false, nil
	default:
		return false, errors.Errorf("converting the specified is adjusted to UTC value '%s' to bool", s)
	}
}

func timeUnitFromString(s string) (*parquet.TimeUnit, error) {
	tu := parquet.NewTimeUnit()
	switch s {
	case "MILLIS":
		tu.MILLIS = parquet.NewMilliSeconds()
	case "MICROS":
		tu.MICROS = parquet.NewMicroSeconds()
	case "NANOS", "":
		tu.NANOS = parquet.NewNanoSeconds()
	default:
		return nil, errors.Errorf("unsupported time unit '%s' specified", s)
	}

	return tu, nil
}

func isTypeCompatible(pt *parquet.Type, gt reflect.Type) bool {
	if gt.Kind() == reflect.Ptr {
		gt = gt.Elem()
	}

	switch *pt {
	case parquet.Type_INT96:
		return (isByteArray(gt) && gt.Len() == 12)
	default:
		return true
	}
}

func isLogicalTypeCompatible(lt *parquet.LogicalType, gt reflect.Type) bool {
	if gt.Kind() == reflect.Ptr {
		gt = gt.Elem()
	}

	switch {
	case lt.IsSetSTRING() || lt.IsSetENUM() || lt.IsSetJSON() || lt.IsSetBSON():
		return gt.Kind() == reflect.String || isByteSlice(gt)
	case lt.IsSetDECIMAL():
		return gt.Kind() == reflect.Int32 || gt.Kind() == reflect.Int64 || isByteSlice(gt) || isByteArray(gt)
	case lt.IsSetDATE():
		return gt.Kind() == reflect.Int32 || isGoTime(gt)
	case lt.IsSetTIME():
		return (gt.Kind() == reflect.Int32 && lt.TIME.Unit.MILLIS != nil) || (gt.Kind() == reflect.Int64 && lt.TIME.Unit.MILLIS == nil) || isGoParquetTime(gt)
	case lt.IsSetTIMESTAMP():
		return gt.Kind() == reflect.Int64 || isGoTime(gt)
	case lt.IsSetUUID():
		return isByteArray(gt) && gt.Len() == 16
	default:
		return false
	}
}

func isByteSlice(gt reflect.Type) bool {
	return gt.Kind() == reflect.Slice && gt.Elem().Kind() == reflect.Uint8
}

func isByteArray(gt reflect.Type) bool {
	return gt.Kind() == reflect.Array && gt.Elem().Kind() == reflect.Uint8
}

func isGoTime(gt reflect.Type) bool {
	return gt.Kind() == reflect.Struct && gt.ConvertibleTo(reflect.TypeOf(time.Time{}))
}

func isGoParquetTime(gt reflect.Type) bool {
	return gt.Kind() == reflect.Struct && gt.ConvertibleTo(reflect.TypeOf(goparquet.Time{}))
}
